File: ios/MagicFormsMobile/Services/SignatureAPI.swift
import Foundation

enum SignatureAPI {
    private static let decoder = JSONDecoder()

    static func fetchSignatures(token: String) async throws -> SignaturesListResponse {
        let url = APIConfig.url(path: "/api/v1/signatures/")
        let data = try await AuthAPI.getData(url: url, token: token)
        let response = try decoder.decode(SignaturesListResponse.self, from: data)
        if response.ok { return response }
        throw try AuthAPI.decodeError(from: data)
    }

    static func uploadSignature(
        token: String,
        pngData: Data,
        label: String,
        replaceSignatureId: Int?
    ) async throws -> SignatureMutationResponse {
        let path: String
        if let replaceSignatureId {
            path = "/api/v1/signatures/\(replaceSignatureId)/image/"
        } else {
            path = "/api/v1/signatures/"
        }
        let url = APIConfig.url(path: path)
        let data = try await AuthAPI.uploadMultipart(
            url: url,
            token: token,
            fields: label.isEmpty ? [:] : ["label": label],
            fileField: "image",
            fileData: pngData,
            fileName: "signature.png",
            mimeType: "image/png"
        )
        let response = try decoder.decode(SignatureMutationResponse.self, from: data)
        if response.ok { return response }
        throw try AuthAPI.decodeError(from: data)
    }

    static func setPrimary(token: String, signatureId: Int) async throws {
        let url = APIConfig.url(path: "/api/v1/signatures/\(signatureId)/primary/")
        let data = try await AuthAPI.postData(url: url, body: Data("{}".utf8), token: token)
        if let body = try? decoder.decode(SignatureMutationResponse.self, from: data), body.ok {
            return
        }
        throw try AuthAPI.decodeError(from: data)
    }

    static func deleteSignature(token: String, signatureId: Int) async throws {
        let url = APIConfig.url(path: "/api/v1/signatures/\(signatureId)/")
        let data = try await AuthAPI.deleteData(url: url, token: token)
        if let body = try? decoder.decode(SignatureMutationResponse.self, from: data), body.ok {
            return
        }
        throw try AuthAPI.decodeError(from: data)
    }
}

File: android/app/src/main/java/com/magicforms/mobile/data/api/SignatureApi.kt
package com.magicforms.mobile.data.api

import com.magicforms.mobile.data.models.SignatureMutationResponse
import com.magicforms.mobile.data.models.SignaturesListResponse
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class SignatureApi {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val authApi = AuthApi()
    private val pngMedia = "image/png".toMediaType()

    fun fetchSignatures(token: String): SignaturesListResponse {
        val data = authApi.get(ApiConfig.url("/api/v1/signatures/"), token)
        return json.decodeFromString(data)
    }

    fun uploadSignature(
        token: String,
        pngBytes: ByteArray,
        label: String,
        replaceSignatureId: Int?,
    ): SignatureMutationResponse {
        val url = if (replaceSignatureId != null) {
            ApiConfig.url("/api/v1/signatures/$replaceSignatureId/image/")
        } else {
            ApiConfig.url("/api/v1/signatures/")
        }
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .apply {
                if (label.isNotBlank()) {
                    addFormDataPart("label", label)
                }
                addFormDataPart(
                    "image",
                    "signature.png",
                    pngBytes.toRequestBody(pngMedia),
                )
            }
            .build()
        val data = authApi.postMultipart(url, body, token)
        return json.decodeFromString(data)
    }

    fun setPrimary(token: String, signatureId: Int) {
        authApi.post(ApiConfig.url("/api/v1/signatures/$signatureId/primary/"), "{}", token)
    }

    fun deleteSignature(token: String, signatureId: Int) {
        authApi.delete(ApiConfig.url("/api/v1/signatures/$signatureId/"), token)
    }
}
